src/proj.h

```diff
@@ -0,0 +1,205 @@
+    in connection with *but before* projects.h (the latter may be needed in
+    some cases, where it is necessary to access this "bare essentials" API,
+    while still having direct access to PJ object internals)
+******************************************************************************/
+
+/* Data type for generic geodetic observations */
+struct OBSERVATION;
+typedef struct OBSERVATION OBSERVATION;
+
+/* Data type for generic geodetic 3D data */
+union PJ_TRIPLET;
+typedef union PJ_TRIPLET PJ_TRIPLET;
+
+/* Data type for generic geodetic 3D data plus epoch information */
+union PJ_SPATIOTEMPORAL;
+typedef union PJ_SPATIOTEMPORAL PJ_SPATIOTEMPORAL;
+
+#ifndef PROJECTS_H
+/* Data type for projection/transformation information */
+struct PJconsts;
+typedef struct PJconsts PJ;         /* the PJ object herself */
+
+/* Omega, Phi, Kappa: Rotations */
+typedef struct {double o, p, k;}  OPK;
+
+/* Easting, Northing, and some kind of height (orthometric or ellipsoidal) */
+typedef struct {double e, n, h;}  ENH;
+
+/* Geodetic spatiotemporal coordinate types */
+typedef struct { double   x,   y,  z, t; }  XYZT;
+typedef struct { double   e,   n,  h, t; }  ENHT;
+typedef struct { double   u,   v,  w, t; }  UVWT;
+typedef struct { double lam, phi,  z, t; }  LPZT;
+
+/* Classic proj.4 pair/triplet types */
+typedef struct { double   u,   v; }  UV;
+typedef struct { double   x,   y; }  XY;
+typedef struct { double lam, phi; }  LP;
+
+typedef struct { double   x,   y,  z; }  XYZ;
+typedef struct { double   u,   v,  w; }  UVW;
+typedef struct { double lam, phi,  z; }  LPZ;
+
+/* Ancillary pairs and triplets for geodetic computations */
+
+/* Degrees, minutes, and seconds */
+typedef struct { double d, m,  s; }  DMS;
+
+/* Geoid undulation and deflections of the vertical */
+typedef struct { double   N,   z,  e; }  PJ_ANC_GEOIDAL;
+
+/* Ellipsoidal parameters */
+typedef struct { double   a,   f; }  PJ_ANC_ELLIPSOIDAL;
+#endif  /* ndef PROJECTS_H */
+
+union PJ_SPATIOTEMPORAL {
+#ifndef PROJECTS_H
+    XYZT xyzt;
+    UVWT uvwt;
+    ENHT enht;
+    LPZT lpzt;
+    ENH  enh;
+#endif
+    double v[4]; /* Who cares - it's just a vector! */
+    XYZ  xyz;
+    UVW  uvw;
+    LPZ  lpz;
+    XY   xy;
+    UV   uv;
+    LP   lp;
+};
+
+
+
+/* Avoid preprocessor renaming and implicit type-punning: Use a union to make it explicit */
+union PJ_TRIPLET {
+#ifndef PROJECTS_H
+    OPK    opk;
+    ENH    enh;
+    PJ_ANC_GEOIDAL     nze;
+    PJ_ANC_ELLIPSOIDAL ell;
+#endif
+    double v[3]; /* Who cares - it's just a vector! */
+    XYZ    xyz;
+    LPZ    lpz;
+    UVW    uvw;
+    XY     xy;
+    LP     lp;
+    UV     uv;
+};
+
+struct OBSERVATION {
+    PJ_SPATIOTEMPORAL coo;
+    PJ_TRIPLET anc;
+    int id;
+    unsigned int flags;
+};
+
+
+#ifndef PROJECTS_H
+/* Apply transformation to observation - in forward or inverse direction */
+OBSERVATION pj_apply (PJ *P, int direction, OBSERVATION obs);
```

Comment:
@QuLogic mentioned the same thing. This is my reply to him:

I'm slightly undecided on this... I introduced the enum in response to a comment by @kbevers who felt the need for a linguistic, rather than mathematical, reminder of "what is what".

My own feeling is that having the "direction" argument directly indicate the (mathematically interpreted) power of the operator (1 forward, -1 inverse) would make ample sense. And since in geodesy numerical stability is important, I would like to have an additional version of the roundtrip test, where e.g. direction=2000 would mean "deviation from input observation for a series of 2000 roundtrips".

Hence, I would prefer to keep the direction argument an int, for future use